### PR TITLE
Dual Stack VIPs for on-prem deployments

### PR DESCRIPTION
**Pull request by cybertron**
Originally the on-prem loadbalancer architecture supported only single-stack
ipv4 or ipv6 deployments. When dual stack support was added for the baremetal
platform, we carried over the VIP configuration that only allowed for a single
VIP to be specified. The thinking was that in dual stack clusters every node
would have access to either ipv4 or ipv6 endpoints. While this is true of the
nodes, it is not necessarily true of the applications that run in the cluster
and may need to interact with the API or ingress services.

This design is proposing the addition of a second VIP for dual stack clusters
so API and ingress are accessible from either ip version.

### CODE REVIEW

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cgwalters) [inline enhancements/network/on-prem-dual-stack-vips.md:248]: Hmm, I don't think the MCO should be mutating the infrastructure object.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cgwalters) [inline enhancements/network/on-prem-dual-stack-vips.md:250]: Right yes, the controllerconfig is just an internal implementation detail of the MCO we can change at will.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cgwalters) [inline enhancements/network/on-prem-dual-stack-vips.md:142]: I like this option the best.  I'd say though it's probably good if this is a relatively "quiet" deprecation at least to start; the cost of carrying the auto-detection code seems low.

**Review by cgwalters — APPROVED** [review]
Very well written enhancement, overall LGTM!  Just some minor things.

(Feel free to at least skim https://github.com/openshift/enhancements/pull/1032 and review/approve in trade for this review :wink: )

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:142]: Yeah, I'm looking at this deprecation more as a "don't use this anymore" than "this is going away". From previous discussions with the installer team I'm not sure we're supposed to break old install-configs anyway, so we may have to carry it indefinitely.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:248]: It certainly doesn't seem right. If there's somewhere more appropriate that we could put migration code I'd be open to that, but if not I think we can live with a slight mismatch between what's in infrastructure and what's in controllerconfig.

Comment (kikisdeliveryservice) [reply enhancements/network/on-prem-dual-stack-vips.md:248]: I think you'd be looking to make changes here in createDiscoveredControllerConfigSpec() to make sure that the correct fields are populated in the controller config: https://github.com/openshift/machine-config-operator/blob/61159678d9bf051a5e8a017210a349f8c643b910/pkg/operator/render.go#L111

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (kikisdeliveryservice) [reply enhancements/network/on-prem-dual-stack-vips.md:142]: +1 Agree with Colin. This also kind of sounds like the Platform/PlatformStatus thing we had to do early on. Where old clusters had an empty new field and we had to populate it correctly for born-in 4.1-ish clusters so they wouldnt break. Overall, we can't break people but it's not the first time we had to do something like this so it should be ok.

Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:142]: Cool, this is the one I would tend to prefer as well since it matches what I'm proposing for the api side, and it's good to hear we're all in agreement so far.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:248]: Thanks, I'll add that link to this section as well.

**Review by cybertron — COMMENTED** [review]
Just a note that I have fixed the lint errors locally, but I'm holding off on pushing a new version of this to avoid orphaning comments. I'll push the cleaned up version after I get someone from the api and installer teams to look at this.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (aojea) [inline enhancements/network/on-prem-dual-stack-vips.md:37]: cc @danwinship for API server dual-stack https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2438-dual-stack-apiserver

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (danwinship) [inline enhancements/network/on-prem-dual-stack-vips.md:242]: FTR, the apiserver situation is slightly more complicated than this: it doesn't support having the `kubernetes.default` Service be dual-stack, and it doesn't have any way of recognizing / talking about dual-stack IPs, but it _can_ take advantage of Linux's support for dual-stack listening sockets. Currently cluster-kube-apiserver-operator explicitly passes `bindNetwork: tcp4` or `bindNetwork: tcp6` in the kube-apiserver config to prevent that, but if we changed it to pass `bindNetwork: tcp` instead, then we'd get an apiserver that accepted connections on both the IPv4 and IPv6 IPs of its node (even though it would only create `Endpoints` for one family or the other).

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (danwinship) [inline enhancements/network/on-prem-dual-stack-vips.md:235]: So if you're going to be fiddling with all this stuff anyway [we may be able to fix UPI node IP detection too](https://bugzilla.redhat.com/show_bug.cgi?id=1949827#c59):

> Currently, for UPI, if the MCO knows the apiserver IP at install time, it will write it into the nodeip-configuration service, and so we will pick the IP with the most direct route to the apiserver (just like in the IPI case). However, in some cases (notably vSphere UPI), the MCO does not know the apiserver IP at install time and thus cannot do this. From what I can tell though, the apiserver IP is still *known* (eg, to the installer) in this case, it's just not recorded anywhere that the MCO can see at install time. So if we fix the install-time config plumbing so that the MCO always has access to the configured apiserver IP, then it should always be able to pass that to nodeip-configuration, and we should always get the right IP. (So this probably requires one or more additions to openshift/api, plus changes to openshift/installer to fill in the new API, and changes to openshift/machine-config-operator to consume the new API. It should not require any changes to openshift/baremetal-runtimecfg.)

I haven't dug too deeply into this but it seems like while we may need to continue having platform-specific `spec` VIP fields, we might be able to have platform-_independent_ `status` VIP fields? (Which would really simplify [this](https://github.com/openshift/machine-config-operator/blob/bbe78cfe1eb9bfa9dc5e8fe27fb8155e648dd848/pkg/operator/render.go#L258).)

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:161]: Hmm, so I was envisioning the pseudo-config template to have a conditional second VIP:
```
{{ if Cluster.APIVIPv4 }}
vrrp_instance apiv4 {
  virtual_ipaddress { {{ Cluster.APIVIPv4 }} }
  unicast_peers {
    {{- range .LBConfig.Backendsv4 -}}
    {{.Address}}
    {{- end}}
  }
}
{{ end }}
{{ if Cluster.APIVIPv6 }}
vrrp_instance apiv6 {
  # Same thing but s/v4/v6/
}
```
But we should be able to do the same thing as a loop, as long as we can keep the VIP list and backend list consistent with each other. I'm not exactly sure how to iterate lists in parallel in a Go template, but I assume there's some way to do it. If not we may have to modify the runtimecfg template data structure so there's a single top-level list.

Not having specific v4 and v6 VIP fields would also fix the UX issue you pointed out where the user needs to make sure to set the right field in single stack deployments.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:194]: If we're using a list for install-config, I wonder if we just make these lists too. If we're going to use the secondary terminology for the extra VIP we might as well do the same in install-config and then we don't have to deprecate any fields, which might be a point in favor of this approach. Maybe I'll write these two options up as alternatives and see if that shakes out any details that would make one clearly better than the other.

We have the flexibility to mix and match data structures from install-config -> api -> render data, but it's simpler if we avoid bouncing between discrete fields and lists.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:242]: Yeah, this was a bit hand-wavy because I couldn't remember exactly what you said was missing from dual stack in the apiserver. I'll rewrite this to be more accurate.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:235]: Hmm, that may be possible. I'm not familiar with UPI installer data, but if it's just adding a field somewhere it's probably doable.

**Review by cybertron — COMMENTED** [review]
Thanks for all the discussion! I'll try to get a new revision of this pushed tomorrow that incorporates the feedback so far.

**Comment by cybertron** [comment]
Okay, after some more mulling I think the primary/secondary arrangement is going to be best for both user experience and ease of implementation. Not having to deprecate any fields limits the impact of this on users, and also removes the need to migrate any values from existing api records. We can always convert the two fields into a list for processing if we need to.

I've updated the enhancement to reflect this, and left the other options as subsections in the alternatives section for reference.

@danwinship @kikisdeliveryservice @cgwalters please take another look and let me know if you agree with my conclusions.

**Review by danwinship — COMMENTED** [review]
seems simple. one question:

**Comment by kikisdeliveryservice** [comment]
From an MCO perspective this LGTM

/lgtm

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (aojea) [inline enhancements/network/on-prem-dual-stack-vips.md:195]: does this mean that haproxy does NAT64?

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (kirankt) [inline enhancements/network/on-prem-dual-stack-vips.md:306]: This would be my pick for the installer API change. The fields are unambiguous to the end user. Granted that this option is a little more intensive to implement, it shouldn't be too complicated. We can start using the new v4/v6 fields right away and add the correct default (for ipv4, ipv4 or dualstack) to the deprecated fields.

**Review by kirankt — COMMENTED** [review]
Overall the proposal is clear and the broad idea is fine by me. Although I vote for alternate 2 design for install config changes. I'd like to see input from @patrickdillon and @staebler as well. Thanks

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (romfreiman) [inline enhancements/network/on-prem-dual-stack-vips.md:63]: Another significant story is hub/spoke mode where the spoke is ipv6 only

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (romfreiman) [inline enhancements/network/on-prem-dual-stack-vips.md:205]: Lets add a section here how will it affect assisted installer / ZTP. Once of the main goals of this work is to allow ZTP deployments of similar topologies

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (mkowalski) [inline enhancements/network/on-prem-dual-stack-vips.md:325]: I'm wondering if this data structure wouldn't be the most in-line with other network parameters configurable nowadays. We have already `clusterNetworks`, `serviceNetworks` and so on that are plural, what does not necessarily mean that they accept every single combination.

When in Assisted Installer we were implementing dual-stack support, we had similar concerns - whether fields that in single-stack accept 1 entry and in dual-stack exactly 2 entries deserve to be lists or should we take the simplest path by introducing `-secondary` suffix. We picked the former as a way of keeping the configuration elegant for the price of internal validations to be a bit more complex.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:195]: I'm not sure how it works, I just know that it did work when we added v6 vips with https://github.com/yboaron/kepalived-v6

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:325]: Dan mentioned the same thing in his comments on a previous revision and if we were doing a clean sheet design I'd absolutely prefer this. With the existing config though, I think there's also an elegance to not having multiple fields that configure the same thing. Yes, we can handle that, but is this enough better to justify the complication of deprecating the old field and complicating the backend implementation? I don't think so, but I don't have strong feelings about this. Most of the options presented here are perfectly valid (except the one Kiran noted above) so I'm open to the alternatives if others feel strongly about it.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:306]: This was the frontrunner initially, but Dan pointed out that if we're going to deprecate the old fields it would be more consistent with the other network settings to make the new fields lists (as in the next alternative).

That said, I'm not opposed to any of these options (except the one you noted above), so if the installer team prefers this I'm not inclined to argue too strenuously. :-)

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:63]: Is there a meaningful difference between "spoke" and "application"?

**Review by cybertron — COMMENTED** [review]
Okay, I think the diversity of opinions here shows how most of the alternatives would work fine. Each option has it benefits and drawbacks, so I'm not dying on any of these hills. However, we do need to pick one that everyone finds acceptable so we can progress with this work.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (sadasu) [inline enhancements/network/on-prem-dual-stack-vips.md:128]: Can we call these new configs apiVIPv6 and ingressVIPv6 instead of calling them "Secondary" ? 

If it is necessary to mark one of them as primary & the other secondary, we should have another optional config like primaryApiVIP with a default value of apiVIP. Assuming the v6 VIPs as secondary seems arbitrary.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (romfreiman) [reply enhancements/network/on-prem-dual-stack-vips.md:63]: Different cluster. Different user story. But not siginificant

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (romfreiman) [reply enhancements/network/on-prem-dual-stack-vips.md:205]: The main goal of this enhancement is to cover hub/spoke scenario AFAIK. Im not saying it should be covered by this enhancement, but it should be considered.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (sadasu) [reply enhancements/network/on-prem-dual-stack-vips.md:281]: This alternative seems like it is the most unambiguous to the user. 

While describing the preferred solution, you mentioned that it would be beneficial to mark one address as the primary (and the other as secondary). So, in that case, for this alternative, would we also need another optional config item to point to the promaryVIP? For example, primaryApiVIP could have the values apiVIP or apiVIPv6. If unspecified it defaults to apiVIP.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (sadasu) [reply enhancements/network/on-prem-dual-stack-vips.md:306]: Most unambiguous. Again we need a way for the user to indicate what the primary VIP is if that is even required.

**Comment by sadasu** [comment]
Thanks for the really nice explanation around each of the options and their pro and cons. My vote goes to the use of plurals or list in the install config and api because it appears to be future proof (in case there is ever a case where we would need more than 1 v4 and 1 v6 address).

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (sadasu) [inline enhancements/network/on-prem-dual-stack-vips.md:352]: The preferred solution points to the need to have a primary VIP and how that becomes easy with that solution. This solution with plurals/list has a similar advantage if we instruct the user to list the primary address as the 1st in the list.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (aojea) [reply enhancements/network/on-prem-dual-stack-vips.md:195]: ah, cool, it was mostly curiosity

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (mhrivnak) [inline enhancements/network/on-prem-dual-stack-vips.md:347]: I could imagine an on-prem scenario where there is both an application network plus an independent management network. Since we blur the lines of the k8s API, using it both for infrastructure management and application management, it would be necessary to have the API service listening on both networks.

Likewise with RHACM you could imagine a hub-and-spoke model where the hub (management) cluster has special access to multiple otherwise-separate networks so it can manage infrastructure in multiple locations / facilities / regions etc. It would benefit from having VIPs on those multiple networks.

I don't know that there's been any direct demand for the above patterns, but they seem plausible.

**Review by mhrivnak — CHANGES_REQUESTED** [review]
Since this will require similar API changes in assisted-service, whose API is an extension of hive's API and is a supported method for installing openshift, the assisted-service team would like to decide together what the API change should look like in both places. I think it makes sense to consider the assisted-service changes as an integral part of this enhancement.

**Comment by sadasu** [comment]
I would like to propose an alternative that actually combines 2 proposals in the enhancement. Keep the existing apiVIP in the install_config and add **additionalApiVIPs** as a list. That has all the advantages listed for preferred solution with the additional advantage of easy extensibility.

 ```yaml
platform:
  baremetal:
    apiVIP: 192.168.111.5
    # Optional
    additionalApiVIPs:
    - f00::5
```

Changes to the Infrastructure API are again a combination what has already been detailed in the 2 sections.

**Review by cybertron — COMMENTED** [review]
Oh, my inline comments didn't get posted. Here you go!

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (mhrivnak) [reply enhancements/network/on-prem-dual-stack-vips.md:50]: I see. So this is referring to a single-stack application running outside the cluster that needs to interact with the api or ingress? I read it as a single-stack application running in the cluster.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (staebler) [inline enhancements/network/on-prem-dual-stack-vips.md:304]: There should not be any logic needed in machine-config-operator or baremetal-runtimecfg to handle older cluster configs. The installer would need to populate the new fields with the values from the deprecated fields. The backend should be able to rely on the new fields always being used.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (staebler) [inline enhancements/network/on-prem-dual-stack-vips.md:291]: This is not a problem, as far as I can tell.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (staebler) [inline enhancements/network/on-prem-dual-stack-vips.md:298]: That validation, or some variant of it, is needed regardless of which option is selected.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (staebler) [reply enhancements/network/on-prem-dual-stack-vips.md:325]: This alternative is my preference. It keeps it consistent with other fields. Deprecation is not terribly complex to achieve. As discussed above, the backend implementation should not be effected by deprecated fields.

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cgwalters) [reply enhancements/network/on-prem-dual-stack-vips.md:304]: MCO doesn't parse the install config for day 2 operations.

Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:304]: This might be a terminology problem. When I say "cluster config", I'm referring to the Infrastructure and ControllerConfig objects that are populated with the values specified in install-config. I will be more explicit in the next revision.

The main thing is that if we deprecate a field in install-config, that will require deprecation of an API field and migration of the values on the backend as well. We can do that, but it is a small drawback to some of the options.

Comment (staebler) [reply enhancements/network/on-prem-dual-stack-vips.md:304]: OK. It was confusing to include cluster config then in the discussion of the merits of alternative designs for the install config. I view the layout of the install-config fields and the layout of the config.openshift.io API fields as separate concerns. If there is nothing using the install-config in-cluster, then the backend arguments has even less bearing as criteria for which install-config alternative to select.

**Comment by cybertron** [comment]
I've pushed a new revision that changes the implementation to be based on a list of VIPs and addresses some other comments.

**Comment by kikisdeliveryservice** [comment]
Since last review, I don't think anything has changed for us on the MCO end, later revisions seem fine, so still lgtm

File: enhancements/network/on-prem-dual-stack-vips.md
Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:109]: Given that we will need to have backward compatibility code in the installer and MCO to handle existing install-configs and clusters, is this a blocker?

I ask because this is already at high risk of missing 4.11 and we're not likely to be able to absorb a significant amount of additional work. If we could land the underlying feature in 4.11 and wire it in to assisted in 4.12 that would help a lot.

Comment (mkowalski) [reply enhancements/network/on-prem-dual-stack-vips.md:109]: No no, this is not a blocker.

Also, all the work I described above will be handled by the Assisted Installer team, so on your side no additional work would be needed. I just wanted to make sure that changes required in Assisted are written down somewhere at the moment of accepting this change of the structure.

Assisted Installer releases with ACM and not with OCP, so our freezes are usually around a month later than OCPs. So if you deliver for 4.11, Assisted will catch-up around ACM 2.6

Comment (cybertron) [reply enhancements/network/on-prem-dual-stack-vips.md:109]: Excellent, thank you! I will get this added.

**Comment by cybertron** [comment]
Added implementation details for assisted-service based on the earlier discussion. I think this should be ready to go now.

**Comment by cybertron** [comment]
@danwinship Since you're an approver in this repo could you also do that? We have agreement on the design from all of the affected teams now. Thanks.

**Pull request merged**
